src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutor.java

```diff
@@ -152,6 +160,9 @@ public ExecuteResponse executeRemotely(ExecuteRequest request)
                   Operation o = replies.next();
                   operation.set(o);
                   waitExecution.set(!operation.get().getDone());
+                  if (receiver != null) {
```

Comment:
Please add a comment here why you're calling this from here.

Reply:
Basically, I want to have a way to update caller the remote execution status. So call `receiver.onNextOperation` here after getting a reply from server.

Reply:
Sorry for being unclear. I would like to see a comment explaining the problem and how this is solving it. As a reader of this code, I'd wonder why this is here rather than before the execute call, and that's specifically to avoid showing an action as executing when it's still queued locally.

Reply:
> rather than before the execute call

This is the behavior Bazel currently have and causing the problem (showing an action as executing when it's still queued locally).

If we report EXECUTING before the [execute](https://github.com/bazelbuild/bazel/blob/52c8773c3ccbcc9051427d6f86a9fe23217f3977/src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutor.java#L148) call, the action might be queued locally waiting for the connection.

By reporting after got a reply from service, we can assume that the connection is up and service will execute the action as soon as possible. However, this assumption might cause other problems if service:
1. only send a reply after the action is done which will leave Bazel showing action as preparing when it's is executing;
2. or send a reply and queue the action. (Can use metadata to tell the status as explained above to avoid this issue)

Reply:
Ulf means that you should a comment to the code so that in the future people understand why the code is written in this way.

Reply:
Ops.